File: app/src/main/java/bf/fasobizness/bafatech/activities/recrutement/ActivityRecrutementRecherche.java
package bf.fasobizness.bafatech.activities.recrutement;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.facebook.shimmer.ShimmerFrameLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import bf.fasobizness.bafatech.R;
import bf.fasobizness.bafatech.adapters.RecrutementAdapter;
import bf.fasobizness.bafatech.helper.RetrofitClient;
import bf.fasobizness.bafatech.interfaces.API;
import bf.fasobizness.bafatech.interfaces.OnItemListener;
import bf.fasobizness.bafatech.models.Recruit;
import bf.fasobizness.bafatech.utils.DatabaseManager;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ActivityRecrutementRecherche extends AppCompatActivity
        implements OnItemListener {
    // private static final String TAG = "ActivityRecherche";
    private LinearLayout layout_ent_offline, layout_busy_system, layout_no_recruit;
    private ShimmerFrameLayout mShimmerViewContainer;
    private RecrutementAdapter mRecrutementAdapter;
    private ArrayList<Recruit.Recrutement> mRecrutements;
    // private RequestQueue requestQueue;
    private TextView resultats;
    private String query;
    private DatabaseManager databaseManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recrutement_recherche);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle("");
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        toolbar.setNavigationOnClickListener(view -> finish());

        EditText ed_rechercher = findViewById(R.id.ed_rechercher);
        resultats = findViewById(R.id.resultats);

        ed_rechercher.setOnEditorActionListener((v, actionId, event) -> {
            boolean handled = false;
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                query = v.getText().toString();
                resultats.setText(R.string.recherche_en_cours);
                InputMethodManager in = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                if (in != null) {
                    in.hideSoftInputFromWindow(ed_rechercher.getWindowToken(), 0);
                }
                handled = true;
                jsonParse();
            }
            return handled;
        });

        // requestQueue = Volley.newRequestQueue(this);
        mRecrutements = new ArrayList<>();
        RecyclerView mRecyclerView = findViewById(R.id.recyclerview_recrutements);
        mRecrutementAdapter = new RecrutementAdapter(getApplicationContext(), mRecrutements);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setAdapter(mRecrutementAdapter);
        mRecrutementAdapter.setOnItemListener(this);
        if (ed_rechercher.getText().toString().isEmpty()) {
            resultats.setText(R.string.que_recherchez_vous);
        }

        layout_ent_offline = findViewById(R.id.layout_ent_offline);
        layout_busy_system = findViewById(R.id.layout_busy_system);
        layout_no_recruit = findViewById(R.id.layout_no_recruit);
        mShimmerViewContainer = findViewById(R.id.shimmer_view_container);
        Button btn_refresh = findViewById(R.id.btn_refresh);
        btn_refresh.setOnClickListener(v -> jsonParse());

        databaseManager = new DatabaseManager(this);
        List<Recruit.Recrutement> recrutementList = databaseManager.getRecruits();
        mRecrutements.addAll(recrutementList);
        mRecrutementAdapter.notifyDataSetChanged();

    }

    private void jsonParse() {
        mRecrutementAdapter.getFilter().filter(query);
        Log.d("Filtering", "Recherche " + query);
    }

    /*private void jsonParse() {
        mRecrutements.clear();
        mRecrutementAdapter.notifyDataSetChanged();
        layout_busy_system.setVisibility(View.GONE);
        layout_ent_offline.setVisibility(View.GONE);
        layout_no_recruit.setVisibility(View.GONE);
        mShimmerViewContainer.setVisibility(View.VISIBLE);
        mRecrutements.clear();
        mRecrutementAdapter.notifyDataSetChanged();

        API api = RetrofitClient.getClient().create(API.class);
        Call<Recruit> call = api.searchRecruits(query);
        call.enqueue(new Callback<Recruit>() {
            @Override
            public void onResponse(@NonNull Call<Recruit> call, @NonNull Response<Recruit> response) {
                mShimmerViewContainer.setVisibility(View.GONE);
                Recruit recruit = response.body();
                List<Recruit.Recrutement> recrutements = null;
                if (recruit != null) {
                    recrutements = recruit.recrutements;
                }
                if (recrutements != null) {
                    mRecrutements.addAll(recrutements);
                }
                mRecrutementAdapter.notifyDataSetChanged();
                if (mRecrutements.size() == 0) {
                    layout_no_recruit.setVisibility(View.VISIBLE);
                } else {
                    resultats.setVisibility(View.VISIBLE);
                    String reslt = getString(R.string.resultats, mRecrutements.size() + "");
                    resultats.setText(reslt);
                }
            }

            @Override
            public void onFailure(@NonNull Call<Recruit> call, @NonNull Throwable t) {
                mShimmerViewContainer.setVisibility(View.GONE);
                layout_ent_offline.setVisibility(View.VISIBLE);
            }
        });
    }*/

    @Override
    public void onItemClicked(int position) {
        Recruit.Recrutement recrutement = mRecrutements.get(position);
        Intent intent = new Intent(getApplicationContext(), ActivityDetailsRecrutement.class);
        intent.putExtra("recrutement", recrutement);
        startActivity(intent);
    }
}

File: app/src/main/java/bf/fasobizness/bafatech/activities/recrutement/ActivityRecrutementRecherche.kt
package bf.fasobizness.bafatech.activities.recrutement

import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import bf.fasobizness.bafatech.R
import bf.fasobizness.bafatech.adapters.RecrutementAdapter
import bf.fasobizness.bafatech.helper.RetrofitClient
import bf.fasobizness.bafatech.interfaces.API
import bf.fasobizness.bafatech.interfaces.OnItemListener
import bf.fasobizness.bafatech.models.Recruit
import bf.fasobizness.bafatech.models.Recruit.Recrutement
import bf.fasobizness.bafatech.utils.DatabaseManager
import com.facebook.shimmer.ShimmerFrameLayout
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*


class ActivityRecrutementRecherche : AppCompatActivity(), OnItemListener {
    private lateinit var layoutNoRecruit: LinearLayout
    private lateinit var mRecrutementAdapter: RecrutementAdapter
    private lateinit var mRecrutements: ArrayList<Recrutement>
    private lateinit var resultats: TextView
    private lateinit var query: String
    private lateinit var layoutEntOffline: LinearLayout
    private lateinit var layoutBusySystem: LinearLayout
    private lateinit var mShimmerViewContainer: ShimmerFrameLayout
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recrutement_recherche)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.title = ""
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toolbar.setNavigationOnClickListener { finish() }
        val edRechercher = findViewById<EditText>(R.id.ed_rechercher)
        resultats = findViewById(R.id.resultats)
        edRechercher.setOnEditorActionListener { v: TextView, actionId: Int, _: KeyEvent? ->
            var handled = false
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                query = v.text.toString()
                resultats.setText(R.string.recherche_en_cours)
                val `in` = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                `in`.hideSoftInputFromWindow(edRechercher.windowToken, 0)
                handled = true
                jsonParse()
            }
            handled
        }
        mRecrutements = ArrayList()
        val mRecyclerView = findViewById<RecyclerView>(R.id.recyclerview_recrutements)
        mRecrutementAdapter = RecrutementAdapter(applicationContext, mRecrutements)
        mRecyclerView.layoutManager = LinearLayoutManager(this)
        mRecyclerView.adapter = mRecrutementAdapter
        mRecrutementAdapter.setOnItemListener(this)
        if (edRechercher.text.toString().isEmpty()) {
            resultats.setText(R.string.que_recherchez_vous)
        }
        layoutNoRecruit = findViewById(R.id.layout_no_recruit)
        layoutEntOffline = findViewById(R.id.layout_ent_offline)
        layoutBusySystem = findViewById(R.id.layout_busy_system)
        mShimmerViewContainer = findViewById(R.id.shimmer_view_container)
        val btnRefresh = findViewById<Button>(R.id.btn_refresh)
        btnRefresh.setOnClickListener { jsonParse() }
        /*val databaseManager = DatabaseManager(this)
        val recrutementList = databaseManager.recruits
        mRecrutements.addAll(recrutementList)
        mRecrutementAdapter.notifyDataSetChanged()*/
    }

    private fun jsonParse() {
        /*mRecrutementAdapter.filter.filter(query)
        Handler(Looper.getMainLooper()).postDelayed({
            if (mRecrutements.size == 0) {
                layoutNoRecruit.visibility = View.VISIBLE
            } else {
                resultats.visibility = View.VISIBLE
                val reslt = getString(R.string.resultats, mRecrutementAdapter.itemCount.toString() + "")
                resultats.text = reslt
            }
        }, 200)*/

        mRecrutements.clear()
        mRecrutementAdapter.notifyDataSetChanged()
        layoutBusySystem.visibility = View.GONE
        layoutEntOffline.visibility = View.GONE
        layoutNoRecruit.visibility = View.GONE
        mShimmerViewContainer.visibility = View.VISIBLE
        mRecrutements.clear()
        mRecrutementAdapter.notifyDataSetChanged()

        val api = RetrofitClient.getClient().create(API::class.java)
        val call = api.searchRecruits(query)
        call.enqueue(object : Callback<Recruit?> {
            override fun onResponse(call: Call<Recruit?>, response: Response<Recruit?>) {
                mShimmerViewContainer.visibility = View.GONE
                val recruit = response.body()
                var recrutements: List<Recrutement>? = null
                if (recruit != null) {
                    recrutements = recruit.recrutements
                }
                if (recrutements != null) {
                    mRecrutements.addAll(recrutements)
                }
                mRecrutementAdapter.notifyDataSetChanged()
                if (mRecrutements.size == 0) {
                    layoutNoRecruit.visibility = View.VISIBLE
                } else {
                    resultats.visibility = View.VISIBLE
                    resultats.text = getString(R.string.resultats, mRecrutements.size.toString() + "")
                }
            }

            override fun onFailure(call: Call<Recruit?>, t: Throwable) {
                mShimmerViewContainer.visibility = View.GONE
                layoutEntOffline.visibility = View.VISIBLE
            }
        })
    }

    override fun onItemClicked(position: Int) {
        val recrutementArrayList = mRecrutementAdapter.recruit
        val recrutement = recrutementArrayList[position]
        val intent = Intent(applicationContext, ActivityDetailsRecrutement::class.java)
        intent.putExtra("recrutement", recrutement)
        startActivity(intent)
    }
}